### PR TITLE
Pom refactoring

### PR DESCRIPTION
**Pull request by honnix**
# TL;DR
pom.xml files refactoring

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
It's a bit hard to list what is being done so I tried adding comments in the PR where applicable.

Note that in this PR we are not trying to upgrade any deps or plugin versions unless necessary.

A further upgrade and cleanup can be applied later, with the possibility introducing maven-dependency analyse. 

## Tracking Issue
_NA

## Follow-up issue
_NA_

### CODE REVIEW

File: jflyte-aws/src/test/java/org/flyte/jflyte/aws/S3FileSystemIT.java [hunk 1]
```diff
@@ -36,12 +36,15 @@
 import org.testcontainers.containers.localstack.LocalStackContainer;
 import org.testcontainers.junit.jupiter.Container;
 import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.utility.DockerImageName;
 
 @Testcontainers
 public class S3FileSystemIT {
 
   @Container
-  public final LocalStackContainer localStack = new LocalStackContainer().withServices(S3);
```
Comment (honnix) [inline jflyte-aws/src/test/java/org/flyte/jflyte/aws/S3FileSystemIT.java:44]: Deprecation.

File: pom.xml [hunk 2]
```diff
@@ -98,10 +98,8 @@
     <!-- also remember to bump the version in flytekit-bom module -->
     <spotless.version>2.21.0</spotless.version>
     <spotbugs.excludeFilterFile>spotbugs-exclude.xml</spotbugs.excludeFilterFile>
-    <error_prone.version>2.19.1</error_prone.version>
-    <!-- Using the error-prone-javac is required when running on JDK 8 -->
-    <error_prone.javac.version>9+181-r4173-1</error_prone.javac.version>
```
Comment (honnix) [inline pom.xml:103]: This is not used anymore and we are no longer on Java 8.

File: pom.xml [hunk 3]
```diff
@@ -111,6 +109,9 @@
     <maven.compiler.source>${maven.compiler.release}</maven.compiler.source>
     <maven.compiler.target>${maven.compiler.release}</maven.compiler.target>
     <mockito.version>3.3.3</mockito.version>
+    <scala212.version>2.12.17</scala212.version>
```
Comment (honnix) [inline pom.xml:112]: Manage these in parent so it is clearer to see what versions we are using.

File: pom.xml [hunk 4]
```diff
@@ -388,23 +459,6 @@
               <!-- see https://stackoverflow.com/a/57004351 -->
               <arg>-Xpkginfo:always</arg>
             </compilerArgs>
-            <annotationProcessorPaths>
```
Comment (honnix) [inline pom.xml:391]: I had tons of troubles with errorprone in IntelliJ when using `annotationProcessorPaths`.

File: pom.xml [hunk 5]
```diff
@@ -336,9 +382,34 @@
...
+        <scope>import</scope>
+      </dependency>
     </dependencies>
   </dependencyManagement>
 
+  <dependencies>
+    <dependency>
+      <groupId>com.google.errorprone</groupId>
+      <artifactId>error_prone_core</artifactId>
+      <scope>provided</scope>
```
Comment (honnix) [inline pom.xml:399]: This is effectively the same as using `annotationProcessorPaths`.

File: pom.xml [hunk 6]
```diff
@@ -748,6 +802,50 @@
...
+      </activation>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>com.diffplug.spotless</groupId>
+            <artifactId>spotless-maven-plugin</artifactId>
+            <executions>
+              <execution>
+                <goals>
+                  <goal>apply</goal>
```
Comment (honnix) [inline pom.xml:821]: When compiling locally, we apply.

File: pom.xml [hunk 7]
```diff
@@ -748,6 +802,50 @@
...
+      <id>ci</id>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>com.diffplug.spotless</groupId>
+            <artifactId>spotless-maven-plugin</artifactId>
+            <executions>
+              <execution>
+                <goals>
+                  <goal>check</goal>
```
Comment (honnix) [inline pom.xml:840]: When building in CI, we check.

File: jflyte-aws/pom.xml [hunk 8]
```diff
@@ -29,18 +29,6 @@
   <name>JFlyte - AWS Filesystem extension</name>
   <description>Stages jars in S3.</description>
 
-  <dependencyManagement>
-    <dependencies>
-      <dependency>
-        <groupId>org.testcontainers</groupId>
```
Comment (honnix) [inline jflyte-aws/pom.xml:35]: Now managed centrally in parent.

File: integration-tests/pom.xml [hunk 9]
```diff
@@ -140,25 +116,37 @@
...
-              </includes>
-            </configuration>
-          </execution>
-        </executions>
-      </plugin>
     </plugins>
   </build>
+  <profiles>
+    <profile>
+      <id>it</id>
```
Comment (honnix) [inline integration-tests/pom.xml:123]: IT can be very slow when running on Apple silicon, we can disable that by `mvn clean verify -P \!it`.

File: integration-tests/pom.xml [hunk 10]
```diff
@@ -33,24 +33,6 @@
     <enforcer.skip>true</enforcer.skip>
   </properties>
 
-  <dependencyManagement>
-    <dependencies>
-      <dependency>
```
Comment (honnix) [inline integration-tests/pom.xml:38]: Now managed centrally in parent.

File: flytekit-local-engine/src/test/java/org/flyte/localengine/LocalEngineTest.java [hunk 11]
```diff
@@ -185,6 +185,7 @@ public void testStructWorkflow() {
     assertEquals(expectedOutput, outputs.get("outputStructData"));
   }*/
 
+  @SuppressWarnings("LockOnNonEnclosingClassLiteral")
```
Comment (honnix) [inline flytekit-local-engine/src/test/java/org/flyte/localengine/LocalEngineTest.java:188]: New errorprone rule.

File: flytekit-examples-scala/pom.xml [hunk 12]
```diff
@@ -24,23 +24,47 @@
     <version>0.4.25-SNAPSHOT</version>
   </parent>
 
-  <artifactId>flytekit-examples-scala</artifactId>
+  <artifactId>flytekit-examples-scala_2.13</artifactId>
```
Comment (honnix) [inline flytekit-examples-scala/pom.xml:27]: Renamed to reflect Scala version, as best practice.

File: flyteidl-protos/pom.xml [hunk 13]
```diff
@@ -70,53 +69,11 @@
       </plugin>
       <plugin>
         <artifactId>maven-compiler-plugin</artifactId>
-        <version>3.8.1</version>
         <configuration>
```
Comment (honnix) [inline flyteidl-protos/pom.xml:72]: This is already configured by parent.

File: flyteidl-protos/pom.xml [hunk 14]
```diff
@@ -52,9 +52,8 @@
       <plugin>
         <groupId>org.xolstice.maven.plugins</groupId>
         <artifactId>protobuf-maven-plugin</artifactId>
-        <version>0.6.1</version>
         <configuration>
-          <protocArtifact>com.google.protobuf:protoc:3.21.1:exe:${os.detected.classifier}</protocArtifact>
+          <protocArtifact>com.google.protobuf:protoc:${protobuf.version}:exe:${os.detected.classifier}</protocArtifact>
```
Comment (honnix) [inline flyteidl-protos/pom.xml:56]: Use the version defined by parent.

File: .github/workflows/build.yaml [hunk 15]
```diff
@@ -38,7 +38,7 @@ jobs:
 
       - name: Verify with Maven
         if: ${{ github.ref != 'refs/heads/master' }}
-        run: mvn --batch-mode verify
+        run: mvn --batch-mode verify -Pci
```
Comment (honnix) [inline .github/workflows/build.yaml:41]: Enable this profile for spotless check.

File: .github/workflows/build.yaml [hunk 16]
```diff
@@ -50,8 +50,9 @@ jobs:
           echo "test" | gpg --batch --clearsign --pinentry-mode loopback || true
 
           echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>"
-          mvn --batch-mode deploy
+          mvn --batch-mode deploy -Pci
         env:
+          CI: true
```
Comment (honnix) [inline .github/workflows/build.yaml:55]: Set this to true so `spotless:apply` will not be done during CI.

File: flyteidl-protos/pom.xml [hunk 17]
```diff
@@ -70,53 +69,11 @@
...
-            </path>
-            <path>
-              <groupId>com.google.errorprone</groupId>
-              <artifactId>error_prone_core</artifactId>
-              <version>${error_prone.version}</version>
-            </path>
-          </annotationProcessorPaths>
         </configuration>
       </plugin>
-      <plugin>
```
Comment (honnix) [inline flyteidl-protos/pom.xml:117]: This plugin is now globally enabled.

File: flytekit-api/pom.xml [hunk 18]
```diff
@@ -44,6 +44,7 @@
     <dependency>
       <groupId>com.google.errorprone</groupId>
       <artifactId>error_prone_annotations</artifactId>
+      <scope>provided</scope>
```
Comment (honnix) [inline flytekit-api/pom.xml:47]: Not necessary but following other places.

File: flytekit-examples/pom.xml [hunk 19]
```diff
@@ -100,9 +101,6 @@
           </execution>
         </executions>
       </plugin>
-      <plugin>
```
Comment (honnix) [inline flytekit-examples/pom.xml:103]: This plugin is globally enabled in the parent.

File: flytekit-scala-tests/pom.xml [hunk 20]
```diff
@@ -81,10 +108,27 @@
...
           </execution>
+          <execution>
+            <id>attach-javadocs</id>
+            <goals>
+              <goal>doc-jar</goal>
+            </goals>
+          </execution>
         </executions>
       </plugin>
+      <!-- Disabling the default javadoc plugin - we use scala-maven-plugin instead -->
```
Comment (honnix) [inline flytekit-scala-tests/pom.xml:119]: We have been doing this for another scala module, so we should follow.

File: flytekit-scala_2.12/pom.xml [hunk 21]
```diff
@@ -84,19 +98,30 @@
...
+            <id>attach-javadocs</id>
+            <goals>
+              <goal>jar</goal>
+            </goals>
+            <phase>none</phase>
+          </execution>
+        </executions>
       </plugin>
     </plugins>
-    <sourceDirectory>../flytekit-scala_2.13/src/main/scala</sourceDirectory>
```
Comment (honnix) [inline flytekit-scala_2.12/pom.xml:99]: This is unnecessary.

File: flytekit-scala_2.12/pom.xml [hunk 22]
```diff
@@ -74,7 +87,8 @@
         <artifactId>scala-maven-plugin</artifactId>
         <configuration>
           <sourceDir>../flytekit-scala_2.13/src/main/scala</sourceDir>
-          <scalaVersion>${scala.version}</scalaVersion>
+          <testSourceDir>../flytekit-scala_2.13/src/test/scala</testSourceDir>
```
Comment (honnix) [inline flytekit-scala_2.12/pom.xml:90]: Not needed now but might save us some headache later when needed.

File: flytekit-scala_2.13/pom.xml [hunk 23]
```diff
@@ -88,16 +95,29 @@
...
           </execution>
         </executions>
       </plugin>
+      <!-- Disabling the default javadoc plugin - we use scala-maven-plugin instead -->
       <plugin>
-        <artifactId>maven-jar-plugin</artifactId>
-      </plugin>
-      <plugin>
-        <artifactId>maven-source-plugin</artifactId>
+        <groupId>org.apache.maven.plugins</groupId>
```
Comment (honnix) [inline flytekit-scala_2.13/pom.xml:110]: Same here.

File: integration-tests/pom.xml [hunk 24]
```diff
@@ -74,22 +56,16 @@
...
     </dependency>
 
-    <!-- dependencies for integration-tests -->
+    <!-- dependencies for integration-tests to make sure these modules are built prior to integration-tests -->
     <dependency>
       <groupId>org.flyte</groupId>
       <artifactId>flytekit-examples</artifactId>
       <scope>test</scope>
     </dependency>
-    <dependency>
```
Comment (honnix) [inline integration-tests/pom.xml:83]: We are currently not using scala examples for IT.

File: integration-tests/pom.xml [hunk 25]
```diff
@@ -74,22 +56,16 @@
...
-      <groupId>org.flyte</groupId>
-      <artifactId>flytekit-examples-scala</artifactId>
-      <scope>test</scope>
-    </dependency>
     <dependency>
       <groupId>org.flyte</groupId>
       <artifactId>jflyte</artifactId>
       <scope>test</scope>
-      <optional>true</optional>
```
Comment (honnix) [inline integration-tests/pom.xml:92]: Not exactly sure why it is here. `test` scope plus `optional` doesn't make much sense.

File: jflyte-utils/pom.xml [hunk 26]
```diff
@@ -47,20 +47,6 @@
       <groupId>org.flyte</groupId>
       <artifactId>jflyte-api</artifactId>
     </dependency>
-    <dependency>
```
Comment (honnix) [inline jflyte-utils/pom.xml:50]: Minior cleanup because this is not needed by jflyte-util.

File: jflyte-utils/pom.xml [hunk 27]
```diff
@@ -90,13 +76,6 @@
       <artifactId>netty-tcnative-boringssl-static</artifactId>
     </dependency>
 
-    <!-- runtime -->
-    <dependency>
```
Comment (honnix) [inline jflyte-utils/pom.xml:94]: Same here.

File: jflyte/pom.xml [hunk 28]
```diff
@@ -171,20 +171,10 @@
       <artifactId>memoryfilesystem</artifactId>
       <scope>test</scope>
     </dependency>
-    <!-- ensure there is at least one slf4j implementation in test classpath so that we can test
-         org/slf4j/impl/StaticLoggerBinder.class is only discovered in child class loader -->
-    <dependency>
```
Comment (honnix) [inline jflyte/pom.xml:176]: This is only needed by jflyte-utils.

File: pom.xml [hunk 29]
```diff
@@ -366,7 +442,7 @@
       <plugins>
         <plugin>
           <artifactId>maven-compiler-plugin</artifactId>
-          <version>3.8.1</version>
```
Comment (honnix) [inline pom.xml:369]: 3.8.1 is very old.

File: flyteidl-protos/pom.xml [hunk 30]
```diff
@@ -52,9 +52,8 @@
       <plugin>
         <groupId>org.xolstice.maven.plugins</groupId>
         <artifactId>protobuf-maven-plugin</artifactId>
-        <version>0.6.1</version>
```
Comment (honnix) [inline flyteidl-protos/pom.xml:55]: Now managed by parent.

File: pom.xml [hunk 31]
```diff
@@ -788,23 +897,6 @@
               </execution>
             </executions>
           </plugin>
-          <plugin>
```
Comment (honnix) [inline pom.xml:791]: It doesn't seem to be needed to reconfigure this plugin for release profile.

File: pom.xml [hunk 32]
```diff
@@ -366,12 +424,12 @@
...
         <plugin>
           <artifactId>maven-compiler-plugin</artifactId>
-          <version>3.8.1</version>
+          <version>3.11.0</version>
           <configuration>
             <showWarnings>true</showWarnings>
             <showDeprecation>true</showDeprecation>
             <forceJavacCompilerUse>true</forceJavacCompilerUse>
-            <failOnWarning>true</failOnWarning>
+            <failOnWarning>false</failOnWarning>
```
Comment (honnix) [inline pom.xml:432]: We can no longer fail on warning because of https://github.com/google/auto/blob/0194ed73fd6194e68bb14c25a97ed8f06d3298e4/service/processor/src/main/java/com/google/auto/service/processor/AutoServiceProcessor.java#L254-L265. It will be a mess to annotate every single task in the repository.

Comment (honnix) [reply pom.xml:432]: This check has always been there but not exercise until https://github.com/google/auto/commit/965e8932814ba79804ff8977ad346837df8c9c35#diff-c28f0ce6609b34aeb32ffc9d26613371c57ecf07d96c97ac7c6ed8ee51f16697.

Useing `-Averify=false` in compiler argument is an option, but that requires we configure maven-compiler-plugin individually for any module having `@AutoService`, otherwise we get warning like `The following options were not recognized by any processor: '[verify]'` which again will fail the build because we `failOnWarning`.

File: pom.xml [hunk 33]
```diff
@@ -85,8 +85,18 @@
...
+    After https://github.com/google/auto/commit/965e8932814ba79804ff8977ad346837df8c9c35#diff-c28f0ce6609b34aeb32ffc9d26613371c57ecf07d96c97ac7c6ed8ee51f16697,
+    auto-service started to apply more verifications, one of which is https://github.com/google/auto/blob/0194ed73fd6194e68bb14c25a97ed8f06d3298e4/service/processor/src/main/java/com/google/auto/service/processor/AutoServiceProcessor.java#L254-L265,
+    and that would generate a warning because we use generics everywhere, e.g. SdkRunnableTask. With
+    warning generated, the build would fail because have failOnWarning configured for maven-compiler-plugin.
+    It would be a mess to annotate everywhere to suppress the warning. It is also possible to disable the verification
+    using -Averify=false in compiler argument, but that requires we configure maven-compiler-plugin individually for any
+    module that uses @AutoService, otherwise we get warning like "The following options were not recognized by any processor: '[verify]'",
+    which again will fail the build because we have failOnWarning.
+     -->
+    <auto-service.version>1.0.2</auto-service.version>
```
Comment (honnix) [inline pom.xml:98]: This is the most recently version that has compatible guava deps with auto-value and error-prone. For some reason, maven could pick up a more recently guava for java compiler to use, but when building in IntelliJ we get errors that is clearly due to incompatible guava version.

**Review by andresgomezfrr — APPROVED** [review]
